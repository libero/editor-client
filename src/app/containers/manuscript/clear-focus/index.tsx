import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import * as manuscriptEditorAction from 'app/actions/manuscript-editor.actions';

export const ClearFocus: React.FC<{ className?: string }> = ({ children, className }) => {
  const dispatch = useDispatch();
  const handleClick = useCallback(() => {
    dispatch(manuscriptEditorAction.removeFocusAction());
  }, [dispatch]);

  return (
    <div className={className} onClick={handleClick}>
      {children}
    </div>
  );
};
