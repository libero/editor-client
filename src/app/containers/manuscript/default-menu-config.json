{
  "button-group": [
    "save",
    "undo",
    "redo",
    "pdf"
  ],
  "menus": [
    "headings",
    "typography",
    "insert"
  ],
  "custom-buttons": []
}