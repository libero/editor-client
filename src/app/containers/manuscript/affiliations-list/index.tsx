import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';
import { useAffiliationStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedAffiliationFormDialog } from 'app/containers/affiliation-form-dialog';
import { ComponentWithId } from 'app/types/utility.types';

export const AffiliationsList: React.FC<ComponentWithId> = ({ id }) => {
  const classes = useAffiliationStyles();
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const editAffiliation = useCallback(
    (aff: Affiliation) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedAffiliationFormDialog,
          props: { affiliation: aff },
          title: 'Edit Affiliation'
        })
      );
    },
    [dispatch]
  );

  const addAffiliation = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedAffiliationFormDialog,
        props: {},
        title: 'Add Affiliation'
      })
    );
  }, [dispatch]);

  const renderAffiliation = (aff: Affiliation): React.ReactNode => (
    <li key={aff.id} className={classes.listItem}>
      <div className={classes.orderLabel}>{affiliations.length > 1 ? `(${aff.label})` : undefined}</div>
      <div className={classes.affiliationInfo}>{aff.getDisplayName()}</div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAffiliation.bind(null, aff)}>
        <EditIcon fontSize="small" />
      </IconButton>
    </li>
  );

  return (
    <section>
      <SectionContainer label="Affiliations" id={id}>
        <ul className={classes.root}>{affiliations.map(renderAffiliation)}</ul>
      </SectionContainer>
      <ActionButton
        title="Add Affiliation"
        variant="addEntity"
        onClick={addAffiliation}
        className={classes.addButton}
      />
    </section>
  );
};
