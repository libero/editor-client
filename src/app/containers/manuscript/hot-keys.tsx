import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { isMacOs } from 'react-device-detect';
import { GlobalHotKeys, configure } from 'react-hotkeys';

import { canApplyMarkToSelection, canRedoChanges, canUndoChanges } from '../../selectors/manuscript-editor.selectors';
import * as manuscriptActions from '../../actions/manuscript.actions';

configure({
  ignoreEventsCondition(event) {
    return false;
  }
});

const withMod = (shortcut: string): string => {
  const mod = isMacOs ? 'cmd' : 'ctrl';
  return `${mod}+${shortcut}`;
};

export const HotKeyBindings: React.FC<{}> = () => {
  const dispatch = useDispatch();

  const canUndo = useSelector(canUndoChanges);
  const canRedo = useSelector(canRedoChanges);
  const canApply = useSelector(canApplyMarkToSelection);

  const invokeUndo = useCallback(() => {
    if (canUndo) {
      dispatch(manuscriptActions.undoAction());
    }
  }, [dispatch, canUndo]);

  const invokeRedo = useCallback(() => {
    if (canRedo) {
      dispatch(manuscriptActions.redoAction());
    }
  }, [dispatch, canRedo]);

  const invokeToggleMark = useCallback(
    (mark: string) => () => {
      if (canApply(mark)) {
        dispatch(manuscriptActions.toggleMarkAction(mark));
      }
    },
    [dispatch, canApply]
  );

  const keymap = {
    UNDO: withMod('z'),
    REDO: [withMod('y'), withMod('shift+z')],
    BOLD: withMod('b'),
    ITALIC: withMod('i'),
    UNDERLINE: withMod('u')
  };

  const handlers = {
    UNDO: invokeUndo,
    REDO: invokeRedo,
    BOLD: invokeToggleMark('bold'),
    ITALIC: invokeToggleMark('italic'),
    UNDERLINE: invokeToggleMark('underline')
  };

  return <GlobalHotKeys allowChanges={true} keyMap={keymap} handlers={handlers} />;
};
