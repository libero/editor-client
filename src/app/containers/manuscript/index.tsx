import React from 'react';
import { useSelector } from 'react-redux';
import { Backdrop, CircularProgress } from '@material-ui/core';

import { isManuscriptLoaded } from '../../selectors/manuscript.selectors';
import './styles.scss';
import { ManuscriptToolbar } from './manuscript-toolbar';
import { ManuscriptEditor } from './manuscript-editor';

const renderBackdrop = (): JSX.Element => (
  <Backdrop open={true}>
    <CircularProgress color="inherit" />
  </Backdrop>
);

export const ManuscriptContainer: React.FC = () => {
  const isLoaded = useSelector(isManuscriptLoaded);
  const renderContent = (): JSX.Element => (
    <React.Fragment>
      <ManuscriptToolbar />
      <ManuscriptEditor />
    </React.Fragment>
  );

  return isLoaded ? renderContent() : renderBackdrop();
};
