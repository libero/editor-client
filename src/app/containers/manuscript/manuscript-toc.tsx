import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Drawer, Divider, Hidden, List, ListItem, ListItemText } from '@material-ui/core';
import classNames from 'classnames';

import { useOutlinePanelStyles } from './styles';
import { getArticleInformation, getJournalMeta } from 'app/selectors/manuscript.selectors';
import { ClearFocus } from 'app/containers/manuscript/clear-focus';
import { getManuscriptBodyTOC } from 'app/selectors/manuscript-editor.selectors';
import { TableOfContents, TOCEntry } from 'app/models/manuscript';
import { scrollIntoViewAction } from 'app/actions/manuscript-editor.actions';

export interface ManuscriptTOCProps {
  tocOpen: boolean;
  handleTocToggle(): void;
}

export const ManuscriptTOC: React.FC<ManuscriptTOCProps> = (props) => {
  const classes = useOutlinePanelStyles(props);
  const dispatch = useDispatch();
  const { tocOpen, handleTocToggle } = props;

  const articleInfo = useSelector(getArticleInformation);
  const journalMeta = useSelector(getJournalMeta);
  const manuscriptBodyTOC = useSelector(getManuscriptBodyTOC);
  const toc: TableOfContents = [
    { level: 1, title: 'Title', id: 'title' },
    { level: 1, title: 'Authors', id: 'authors' },
    { level: 1, title: 'Affiliations', id: 'affiliations' },
    { level: 1, title: 'Abstract', id: 'abstract' },
    { level: 1, title: 'Impact statement', id: 'impactStatement' },
    ...manuscriptBodyTOC,
    { level: 1, title: 'Acknowledgements', id: 'acknowledgements' },
    { level: 1, title: 'References', id: 'references' },
    { level: 1, title: 'Author information', id: 'author-details' },
    { level: 1, title: 'Article information', id: 'article-info' },
    { level: 1, title: 'Related articles', id: 'realted-acticles' }
  ];

  const handleTOCClick = useCallback(
    (entry: TOCEntry) => () => {
      dispatch(scrollIntoViewAction(entry));
    },
    [dispatch]
  );

  const drawer = (
    <React.Fragment>
      <ClearFocus>
        <section className={classes.toolbarPlaceholder} aria-hidden="true">
          <h1 className={classes.title}>Libero Editor</h1>
          <h3 className={classes.infoText}>Software: v{process.env.REACT_APP_VERSION}</h3>
          <h3 className={classNames(classes.infoText, classes.dtd)}>JATS: {articleInfo.dtd} DTD</h3>
          <h3 className={classes.infoText}>Type: {articleInfo.articleType}</h3>
        </section>
      </ClearFocus>
      <Divider />
      <List>
        {toc.map((entry, index) => (
          <ListItem button key={index} onClick={handleTOCClick(entry)} style={{ paddingLeft: entry.level * 16 }}>
            <ListItemText
              disableTypography
              classes={{ root: entry.level === 1 ? classes.tocLevel1 : classes.tocLevel2 }}
              primary={entry.title}
            />
          </ListItem>
        ))}
      </List>
      <ClearFocus className={classes.whiteSpace} />
      <section className={classes.journalMeta}>
        <div>{journalMeta.publisherName}</div>
        <div>ISSN: {journalMeta.issn}</div>
      </section>
    </React.Fragment>
  );

  const renderContent = (): JSX.Element => (
    <nav className={classes.drawer} aria-label="mailbox folders">
      <Hidden smUp implementation="css">
        <Drawer
          variant="temporary"
          anchor="left"
          open={tocOpen}
          onClose={handleTocToggle}
          classes={{
            paper: classes.drawerPaper
          }}
          ModalProps={{
            keepMounted: true // Better open performance on mobile.
          }}
        >
          {drawer}
        </Drawer>
      </Hidden>
      <Hidden xsDown implementation="css">
        <Drawer
          classes={{
            paper: classes.drawerPaper
          }}
          variant="permanent"
          open
        >
          {drawer}
        </Drawer>
      </Hidden>
    </nav>
  );

  return renderContent();
};
