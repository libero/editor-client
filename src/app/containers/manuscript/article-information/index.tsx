import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useSelector, useDispatch } from 'react-redux';

import { SectionContainer } from 'app/components/section-container';
import { getArticleInformation } from 'app/selectors/manuscript.selectors';
import { useArticleInformationStyles } from 'app/containers/manuscript/article-information/styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ArticleInfoFormDialog } from 'app/containers/article-info-form-dialog';

export const ArticleInformation: React.FC<{}> = () => {
  const articleInfo = useSelector(getArticleInformation);
  const dispatch = useDispatch();
  const editArticleInfo = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  }, [dispatch]);

  const classes = useArticleInformationStyles();
  return (
    <SectionContainer label="Article information" className={classes.sectionContainer}>
      <section className={classes.infoSection}>
        <div>
          <strong> Article DOI: </strong> {articleInfo.articleDOI}
        </div>
        <div>
          <strong> Publisher ID: </strong> {articleInfo.publisherId}
        </div>
      </section>
      <IconButton classes={{ root: classes.editButton }} onClick={editArticleInfo}>
        <EditIcon fontSize="small" />
      </IconButton>
    </SectionContainer>
  );
};
