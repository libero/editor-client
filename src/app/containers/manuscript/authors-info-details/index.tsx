import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import EditIcon from '@material-ui/icons/Edit';
import { isBoolean, isEqual } from 'lodash';
import Interweave from 'interweave';

import { SectionContainer } from 'app/components/section-container';
import { getAuthorAffiliations, getAuthors } from 'app/selectors/manuscript.selectors';
import { Person } from 'app/models/person';
import {
  useAuthorDetailStyles,
  useAuthorsDetailsListStyles
} from 'app/containers/manuscript/authors-info-details/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { stringifyEditorState } from 'app/utils/view.utils';
import { OrcidIcon } from 'app/assets/icons';
import { ComponentWithId } from 'app/types/utility.types';

export const AuthorsInfoDetails: React.FC<ComponentWithId> = ({ id }) => {
  const authors = useSelector(getAuthors);
  const classes = useAuthorsDetailsListStyles();
  const dispatch = useDispatch();

  const addAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  return (
    <section>
      <SectionContainer label="Author Information" id={id}>
        {authors.map((author) => (
          <AuthorInfo key={author.id} author={author} />
        ))}
      </SectionContainer>
      <ActionButton variant="addEntity" title="Add Author" onClick={addAuthor} className={classes.addAuthorButton} />
    </section>
  );
};

interface AuthorInfoProps {
  author: Person;
}

const AuthorInfo: React.FC<AuthorInfoProps> = React.memo(({ author }) => {
  const classes = useAuthorDetailStyles();
  const dispatch = useDispatch();
  const affiliations = useSelector(getAuthorAffiliations)(author);

  const editAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author },
        title: 'Edit Author'
      })
    );
  }, [author, dispatch]);

  return (
    <section className={classes.root}>
      <div className={classes.authorInfoContainer}>
        <div className={classes.authorInfoLine}>
          <strong> {author.getDisplayName()} </strong>
        </div>
        <div className={classes.authorInfoLine}>
          {author.isCorrespondingAuthor ? 'Corresponding Author: ' : undefined}
          {author.email}
        </div>
        <div className={classes.authorInfoLine}>
          {affiliations.map((aff) => (
            <div key={aff.id}>{aff.getDisplayName()}</div>
          ))}
        </div>
        <div className={classes.authorInfoLine}>
          {author.isAuthenticated ? <OrcidIcon className={classes.orcidIcon} /> : undefined}{' '}
          <a href={`https://orcid.org/${author.orcid}`} target="_blank" rel="noopener noreferrer">
            {author.orcid}
          </a>
        </div>
        {isBoolean(author.hasCompetingInterest) ? (
          <div className={classes.authorInfoLine}>
            {author.hasCompetingInterest
              ? `Competing interest: ${author.competingInterestStatement}`
              : 'No Competing interest declared'}
          </div>
        ) : undefined}
        {author.bio ? (
          <div className={classes.authorBio}>
            <Interweave content={stringifyEditorState(author.bio)} />
          </div>
        ) : undefined}
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAuthor}>
        <EditIcon fontSize="small" />
      </IconButton>
    </section>
  );
}, isEqual);
