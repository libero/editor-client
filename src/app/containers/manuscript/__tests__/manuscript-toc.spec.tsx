import React from 'react';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Hidden, Drawer, ListItem, Divider, List, ListItemText } from '@material-ui/core';

import { getLoadableStateSuccess } from 'app/utils/state.utils';
import { ManuscriptTOC } from 'app/containers/manuscript/manuscript-toc';

jest.mock('@material-ui/core');

describe('<ManuscriptTOC>', () => {
  const mockStore = configureMockStore([]);
  const OLD_VERSION = process.env.REACT_APP_VERSION;

  beforeEach(() => {
    process.env.REACT_APP_VERSION = '0.0.0-test';
    (Hidden as jest.Mock).mockImplementation(({ children }) => <div data-cmp="hidden">{children}</div>);
    (Drawer['render'] as jest.Mock).mockImplementation(({ children }) => <div data-cmp="drawer">{children}</div>);
    (Divider['render'] as jest.Mock).mockImplementation(({ children }) => <div data-cmp="divider">{children}</div>);
    (List['render'] as jest.Mock).mockImplementation(({ children }) => <div data-cmp="list">{children}</div>);
    (ListItem['render'] as jest.Mock).mockImplementation(({ children }) => <div data-cmp="list-item">{children}</div>);
    (ListItemText['render'] as jest.Mock).mockImplementation(({ children }) => (
      <div data-cmp="list-item-text">{children}</div>
    ));
  });

  afterEach(() => {
    process.env.REACT_APP_VERSION = OLD_VERSION;
    (Hidden as jest.Mock).mockReset();
    (Drawer['render'] as jest.Mock).mockReset();
    (Divider['render'] as jest.Mock).mockReset();
    (List['render'] as jest.Mock).mockReset();
    (ListItem['render'] as jest.Mock).mockReset();
    (ListItemText['render'] as jest.Mock).mockReset();
  });

  it('renders', () => {
    const store = mockStore({
      manuscriptEditor: {
        manuscriptBodyTOC: [
          { level: 1, title: 'H1', id: '1' },
          { level: 2, title: 'H2', id: '2' },
          { level: 3, title: 'H2', id: '3' },
          { level: 4, title: 'H2', id: '4' }
        ]
      },
      manuscript: getLoadableStateSuccess({
        past: [{}],
        present: {},
        future: [{}]
      })
    });
    const wrapper = create(
      <Provider store={store}>
        <ManuscriptTOC tocOpen={false} handleTocToggle={() => undefined} />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
