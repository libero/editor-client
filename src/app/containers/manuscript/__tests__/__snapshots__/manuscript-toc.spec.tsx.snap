// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManuscriptTOC> renders 1`] = `
<nav
  aria-label="mailbox folders"
  className="makeStyles-drawer-1"
>
  <div
    data-cmp="hidden"
  >
    <div
      data-cmp="drawer"
    >
      <section
        aria-hidden="true"
        className="makeStyles-toolbarPlaceholder-3"
      >
        <div
          onClick={[Function]}
        >
          <h1
            className="makeStyles-title-7"
          >
            Libero Editor
          </h1>
          <h3
            className="makeStyles-infoText-8"
          >
            Software: v
            0.0.0-test
          </h3>
          <h3
            className="makeStyles-infoText-8 makeStyles-dtd-9"
          >
            JATS: 
             DTD
          </h3>
          <h3
            className="makeStyles-infoText-8"
          >
            Type: 
          </h3>
        </div>
      </section>
      <div
        data-cmp="divider"
      />
      <div
        data-cmp="list"
      >
        <div
          data-cmp="list-item"
        >
          <div
            data-cmp="list-item-text"
          />
        </div>
        <div
          data-cmp="list-item"
        >
          <div
            data-cmp="list-item-text"
          />
        </div>
      </div>
      <div
        className="makeStyles-whiteSpace-4"
        onClick={[Function]}
      />
      <section
        className="makeStyles-journalMeta-6"
      >
        <div />
        <div>
          ISSN: 
        </div>
      </section>
    </div>
  </div>
  <div
    data-cmp="hidden"
  >
    <div
      data-cmp="drawer"
    >
      <section
        aria-hidden="true"
        className="makeStyles-toolbarPlaceholder-3"
      >
        <div
          onClick={[Function]}
        >
          <h1
            className="makeStyles-title-7"
          >
            Libero Editor
          </h1>
          <h3
            className="makeStyles-infoText-8"
          >
            Software: v
            0.0.0-test
          </h3>
          <h3
            className="makeStyles-infoText-8 makeStyles-dtd-9"
          >
            JATS: 
             DTD
          </h3>
          <h3
            className="makeStyles-infoText-8"
          >
            Type: 
          </h3>
        </div>
      </section>
      <div
        data-cmp="divider"
      />
      <div
        data-cmp="list"
      >
        <div
          data-cmp="list-item"
        >
          <div
            data-cmp="list-item-text"
          />
        </div>
        <div
          data-cmp="list-item"
        >
          <div
            data-cmp="list-item-text"
          />
        </div>
      </div>
      <div
        className="makeStyles-whiteSpace-4"
        onClick={[Function]}
      />
      <section
        className="makeStyles-journalMeta-6"
      >
        <div />
        <div>
          ISSN: 
        </div>
      </section>
    </div>
  </div>
</nav>
`;
